test/elbo.jl

```diff
@@ -22,7 +22,29 @@
     rng = MersenneTwister(123456)
     q_f = Normal.(zeros(10), ones(10))
 
-    @testset "$lik" for lik in Base.uniontypes(ApproximateGPs.ScalarLikelihood)
+    analytic_likelihoods = [
+        ExponentialLikelihood, GammaLikelihood, PoissonLikelihood, GaussianLikelihood
+    ]
+
+    @testset "testing all analytic implementations" begin
+        ### Test that we're not missing any analytic implementation in `analytic_likelihoods`!
+        default_quadrature_method_types = Set([
+            m.sig.types[2] for m in methods(ApproximateGPs._default_quadrature)
```

Comment:
Hmm I'm not a huge fan of going about this in this manner. I can see the advantage in principle of a programatic approach, but my prefered way to ensure we have coverage is to ensure that the code / tests are well organised, and to look at the coverage to prevent regressions / check that everything we think ought to be covered is in fact covered.

Additionally, since the `expected_loglik` implementations are now in their own file, I would prefer these tests move to a corresponding test file. We can then have a few tests for the generic methods, and a `@testset` with a `for` loop that iterates through the analytic likelihoods.

Reply:
Coverage might not necessarily tell you whether that likelihood actually had its analytic expectation implementation covered, though, just that it was covered _somewhere_ (could have been an example that tests some other part!). How else would you suggest to ensure that adding an analytic implementation would always be tested ?

Reply:
(:+1:  to moving it to separate file to keep symmetry with src/ though, will do that now)

Reply:
> Coverage might not necessarily tell you whether that likelihood actually had its analytic expectation implementation covered, though, just that it was covered somewhere (could have been an example that tests some other part!). How else would you suggest to ensure that adding an analytic implementation would always be tested ?

That's a fair point. My view is shaped by my experiences with ChainRules etc, in which this is less of an issue because everything is (pretty much exclusively) unit tested / there's no integration testing inside ChainRules itself really.

I'll have another ponder on this. Give me a few mins.

Reply:
Another way to look at this is that this approach also doesn't guarantee that everything is tested -- it would be possible to obtain a false positive if someone introduced a method of `_default_quadrature` that is less specific than an existing one, but is also applicable to a likelihood in the list. For example, if I added a method
```julia
_default_quadarature(::PoissonLikelihood)
```
it'll pass the test, but won't be hit because the more specific method for `PoissonLikelihood{ExpLink}` will be hit with  the current definition of `likelihoods_to_test`.

I guess my more broad point is that I'm not aware of a good way to ensure that all of the code written in a particular file is unit tested in the place you expect it to be.

Reply:
I've changed my test's implementation, it now directly looks at methods of `expected_loglik`, it looks cleaner, and I'm less sure you could easily get it wrong as it is now. Let me know if you're happy to merge it : )

Reply:
I agree that the new implementation is cleaner, and clearly an improvement over what we had before. If it causes problems for any reason in the future we can revisit, but I'm very happy for this to go in as-is.